pkg/compiler/namespace.go

```diff
@@ -0,0 +1,97 @@
+package compiler
+
+import (
+	"github.com/MontFerret/ferret/pkg/runtime/core"
+	"github.com/pkg/errors"
+	"regexp"
+	"strings"
+)
+
+var fnNameValidation = regexp.MustCompile("^[a-zA-Z]+[a-zA-Z0-9_]*(::[a-zA-Z]+[a-zA-Z0-9_]*)*$")
+
+const emptyNS = ""
+const separator = "::"
+
+type NamespaceBuilder struct {
+	funcs core.Functions
+	name  string
+}
+
+func newRootNamespace() *NamespaceBuilder {
+	ns := new(NamespaceBuilder)
+	ns.funcs = make(core.Functions)
+
+	return ns
+}
+
+func newNamespace(funcs core.Functions, name string) *NamespaceBuilder {
+	return &NamespaceBuilder{funcs, strings.ToUpper(name)}
+}
+
+func (rs *NamespaceBuilder) Namespace(name string) core.Namespace {
```

Comment:
why shortcut for `NamespaceBuilder` is `rs` but not `nb` or `builder` for example?

Reply:
Ah, because ``NamespaceBuilder`` was named differently in early iterations.